### PR TITLE
oracle for action similarity

### PR DESCRIPTION
**Pull request by zhouzypaul**

### CODE REVIEW

**Comment by zhouzypaul** [comment]
@camall3n when pinning other q values during training the DQN, the fully trained agent doesn't have the capacity to solve cartpole (during evaluation, it achieve a score of 176, as opposed to 200 when other q values are not pinned).

File: actgen/agents/dqn.py [hunk 1]
```diff
@@ -79,16 +81,33 @@ def _get_epsilon(self, testing=False):
...
+                    action_taken = batch.action[i]
+                    if self.params['oracle']:
+                        # oracle update
+                        num_dup = self.params['duplicate']
+                        original_a = math.floor(action_taken / num_dup)
+                        similar_a = list(range(original_a * num_dup, original_a * num_dup + num_dup))
+                        all_action_targets[i, similar_a] = target
+                    else:
+                        # normal update
+                        all_action_targets[i, action_taken] = target
```
Comment (camall3n) [inline actgen/agents/dqn.py]: I'm imagining this will get more complicated when we want to incorporate the inverse model idea. Can we abstract some of this code into a helper function `_get_action_similarities()`?

I'm imagining the code here would become something like the following:
```python
with torch.no_grad():
    # compute q_targets, vector of size (batch_size,).
    # It's the same thing as 'targets' now, but I'd like to rename it.

    # compute tensor of size (batch_size, n_actions) with values between 0 and 1.
    # 1 if "fully similar"; 0 if "fully different".
    action_similarities = self._get_action_similarities(batch)

q_predictions = self._get_q_predictions(batch)  # (batch_size, n_actions)
td_error = q_targets - q_predictions  # (batch_size, n_actions)
loss = torch.mean(torch.sum(action_similarities * (td_error ** 2), dim=-1), dim=0)
```

This is just a rough sketch; we may need to expand certain dimensions and/or squeeze the results to get the right sizes.

File: actgen/agents/dqn.py [hunk 2]
```diff
@@ -79,16 +81,33 @@ def _get_epsilon(self, testing=False):
...
+                return all_action_targets
         return targets
 
     def _get_q_predictions(self, batch):
-        q_values = self.q(torch.stack(batch.state).float())
-        q_acted = extract(q_values, idx=torch.stack(batch.action).long(), idx_dim=-1)
+        q_values = self.q(torch.stack(batch.state).float())  # (batch_size, n_actions)
```
Comment (camall3n) [inline actgen/agents/dqn.py:115]: I keep seeing you use `float()` and `long()` everywhere. Is there a reason you're doing that?

I wonder if there's a way we can avoid casting in so many different places...

File: actgen/train.py [hunk 3]
```diff
@@ -55,6 +56,8 @@ def parse_args(self):
                             help='Enable test mode for quickly checking configuration works')
         parser.add_argument('--gscore', default=False, action='store_true',
                             help='Calculate the g-score vs time as training proceeds')
+        parser.add_argument('--oracle', default=False, action='store_true',
+                            help='use our expert domain knowledge to perform oracle update')
```
Comment (camall3n) [inline actgen/train.py]: "to perform oracle _action generalization_"

File: actgen/train.py [hunk 4]
```diff
@@ -92,6 +95,9 @@ def setup(self):
         self.env = env
         self.test_env = test_env
 
+        if self.params['oracle']:
+	        self.params['dqn_train_pin_other_q_values'] = True
```
Comment (camall3n) [inline actgen/train.py]: If `dqn_train_pin_other_q_values` is already set to `False`, I'd prefer to throw an exception and tell the user they messed up, rather than silently changing their hyperparams

**Review by camall3n — CHANGES_REQUESTED** [review]

File: actgen/agents/dqn.py (hunk 2 above)
Comment (zhouzypaul) [reply actgen/agents/dqn.py:115]: It's because I saw you always used `float()` in forward-pass calls in the scaffolding of the DQN code. I'm not sure if we need to cast it, because I assume the default dtype in pytorch would be floats. 

Edit: if I remembered wrong, then it's probably because I saw it from someone else's codebase.

Comment (camall3n) [reply actgen/agents/dqn.py:115]: No, that's right. In general whenever you convert from numpy to pytorch, you need to be specific about what type you want the tensors to have. The numpy default is float64, but the pytorch default is float32, and you'll get type errors if you mix and match.

But my question is about whether there's a way to avoid converting so often.

**Comment by zhouzypaul** [comment]
currently, we are using `torch.nn.SmoothL1Loss` as the training loss we are trying to minimize. Just to confirm, we are switching over to the mean(weighted squared TD error) loss as discussed in the meeting? 

(I think the reason we structured our code to do duplicate stuff (adding and subtracting `q_predictions`) is to cater to the signature of the SmoothL1Loss() method)

File: actgen/agents/dqn.py (hunk 2 above)
Comment (zhouzypaul) [reply actgen/agents/dqn.py:115]: Oh I see.

maybe we could do `torch.set_default_dtype(torch.float64)` at the very beginning of the script?

**Comment by camall3n** [comment]
I think either approach will work. Maybe we can start with the original version, since it's more similar to the baseline algorithm we already tried. (You can click the edit history for the post above to recover the original pseudocode.)

File: actgen/agents/dqn.py (hunk 2 above)
Comment (camall3n) [reply actgen/agents/dqn.py:115]: I believe that will only set the default for newly constructed tensors, not tensors that are converted from numpy.

This isn't a big issue. Let's just keep it in the back of our minds going forward.

**Comment by zhouzypaul** [comment]
Yea that's what I was thinking as well! 

The new commits I made yesterday should have addressed all previous problems (except for the dtype conversion between numpy as torch)

**Review by camall3n — APPROVED** [review]
